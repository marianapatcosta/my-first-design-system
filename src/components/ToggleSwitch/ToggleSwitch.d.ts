import React, { ChangeEvent, CSSProperties } from 'react';
export interface ToggleSwitchProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
    /**
     * ToggleSwitch's label
     */
    label?: string;
    /**
     * label to display inside the ToggleSwitch, on the left side
     */
    leftLabel?: string;
    /**
     * label to display inside the ToggleSwitch, on the right side
     */
    rightLabel?: string;
    /**
     * ToggleSwitch's name
     */
    name?: string;
    /**
     * Is ToggleSwitch checked?
     */
    checked: boolean;
    /**
     * Is ToggleSwitch disabled?
     */
    disabled?: boolean;
    /**
     * Change handler
     */
    onChange: (event?: ChangeEvent<HTMLInputElement>) => void;
}
export declare const ToggleSwitch: React.FC<ToggleSwitchProps>;
