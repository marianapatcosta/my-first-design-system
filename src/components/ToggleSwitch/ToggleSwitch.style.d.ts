export declare const StyledToggleSwitch: import("styled-components").StyledComponent<"label", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledToggleSwitchLabel: import("styled-components").StyledComponent<"span", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledToggleSwitchSlider: import("styled-components").StyledComponent<"span", any, {
    checked: boolean;
    disabled?: boolean | undefined;
    label?: string | undefined;
}, never>;
export declare const StyledToggleInput: import("styled-components").StyledComponent<"input", any, {}, never>;
