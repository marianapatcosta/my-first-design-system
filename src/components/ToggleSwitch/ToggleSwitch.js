var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyledToggleSwitch, StyledToggleSwitchLabel, StyledToggleSwitchSlider, StyledToggleInput, } from './ToggleSwitch.style';
export var ToggleSwitch = function (_a) {
    var className = _a.className, style = _a.style, name = _a.name, label = _a.label, leftLabel = _a.leftLabel, rightLabel = _a.rightLabel, checked = _a.checked, disabled = _a.disabled, onChange = _a.onChange;
    return (_jsxs(StyledToggleSwitch, __assign({ className: className, style: style, disabled: disabled }, { children: [_jsx(StyledToggleInput, { type: 'checkbox', name: name || label, checked: checked, disabled: disabled, onChange: onChange }, void 0), !!label && label, _jsxs(StyledToggleSwitchSlider, __assign({ checked: checked, disabled: disabled, label: label, "data-testid": 'toggle-slider' }, { children: [_jsx(StyledToggleSwitchLabel, __assign({ disabled: disabled }, { children: checked && leftLabel }), void 0),
                    _jsx(StyledToggleSwitchLabel, { children: !checked && rightLabel }, void 0)] }), void 0)] }), void 0));
};
ToggleSwitch.defaultProps = {
    id: '',
    className: '',
    style: {},
    name: '',
    label: '',
    leftLabel: '',
    rightLabel: '',
    checked: false,
    disabled: false,
    onChange: function () { return null; },
};
