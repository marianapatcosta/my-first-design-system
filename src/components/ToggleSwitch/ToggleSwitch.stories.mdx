import { Meta } from '@storybook/addon-docs/blocks';
import { ToggleSwitch } from './ToggleSwitch';

<Meta title="Design System|ToggleSwitch" component={ToggleSwitch} />

# ToggleSwitch
Simple ToggleSwitch component, which mimics the checkbox behavior but with different styling

## Usage
Use ToggleSwitch when the value alternates between two values

- Default ToggleSwitch:
  <Story name='Default'>
    <ToggleSwitch checked={false}
      leftLabel='off' 
      rightLabel='on' 
    />
  </Story>

- ToggleSwitch With Label:
  <Story name='WithLabel'>
    <ToggleSwitch checked={false}
      leftLabel='off' 
      rightLabel='on' 
      label='ToggleSwitch' 
    />
  </Story>

## Other important notes
not applicable.
