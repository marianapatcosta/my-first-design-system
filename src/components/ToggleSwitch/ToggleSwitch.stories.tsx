import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { ToggleSwitch, ToggleSwitchProps } from './ToggleSwitch';

export default {
  title: 'ToggleSwitch',
  component: ToggleSwitch,
  // args behavior in each story
  args: { checked: false, leftLabel: 'off', rightLabel: 'on' },
} as Meta;

const Template: Story<ToggleSwitchProps> = (args) => <ToggleSwitch {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const WithLabel = Template.bind({});
WithLabel.args = { label: 'ToggleSwitch' };
