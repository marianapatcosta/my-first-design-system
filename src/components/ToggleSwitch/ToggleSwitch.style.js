var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, disabledStyles, truncateStyles, colors, } from '../../styles';
export var StyledToggleSwitch = styled.label(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  display: flex;\n  align-items: center;\n  height: auto;\n  width: auto;\n  font-size: 95%;\n  color: ", ";\n  font-style: italic;\n  ", ";\n\n  ", "\n"], ["\n  ", "\n  display: flex;\n  align-items: center;\n  height: auto;\n  width: auto;\n  font-size: 95%;\n  color: ", ";\n  font-style: italic;\n  ", ";\n\n  ", "\n"])), baseStyles, colors.highlight, truncateStyles, function (_a) {
    var disabled = _a.disabled;
    return disabled && disabledStyles + " color: " + colors.disabled;
});
export var StyledToggleSwitchLabel = styled.span(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  line-height: 0.75rem;\n  padding: 0 0.125rem;\n  align-self: center;\n  white-space: nowrap;\n  user-select: none;\n  cursor: pointer;\n  font-size: 90%;\n  font-weight: 700;\n\n  ", "\n"], ["\n  line-height: 0.75rem;\n  padding: 0 0.125rem;\n  align-self: center;\n  white-space: nowrap;\n  user-select: none;\n  cursor: pointer;\n  font-size: 90%;\n  font-weight: 700;\n\n  ", "\n"])), function (_a) {
    var disabled = _a.disabled;
    return disabled && "" + disabledStyles;
});
export var StyledToggleSwitchSlider = styled.span(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  justify-content: space-between;\n  align-items: flex-end;\n  border-radius: 2rem;\n  border: 0.063rem solid ", ";\n  cursor: pointer;\n  height: 1rem;\n  width: 2.4rem;\n  line-height: 85%;\n  position: relative;\n  padding: 0.125rem;\n  margin: 0.125rem;\n  transition: 0.4s;\n  -webkit-transition: 0.4s;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  :before {\n    background-color: ", ";\n    opacity: 0.5;\n    border-radius: 50%;\n    content: '';\n    height: 1.25rem;\n    width: 1.25rem;\n    left: -0.063rem;\n    bottom: 0.015rem;\n    position: absolute;\n    transition: 0.4s;\n    -webkit-transition: 0.4s;\n    -webkit-border-radius: 50%;\n    -moz-border-radius: 50%;\n    -ms-border-radius: 50%;\n    -o-border-radius: 50%;\n  }\n\n  ", "\n\n  ", "\n"], ["\n  display: flex;\n  justify-content: space-between;\n  align-items: flex-end;\n  border-radius: 2rem;\n  border: 0.063rem solid ", ";\n  cursor: pointer;\n  height: 1rem;\n  width: 2.4rem;\n  line-height: 85%;\n  position: relative;\n  padding: 0.125rem;\n  margin: 0.125rem;\n  transition: 0.4s;\n  -webkit-transition: 0.4s;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  :before {\n    background-color: ", ";\n    opacity: 0.5;\n    border-radius: 50%;\n    content: '';\n    height: 1.25rem;\n    width: 1.25rem;\n    left: -0.063rem;\n    bottom: 0.015rem;\n    position: absolute;\n    transition: 0.4s;\n    -webkit-transition: 0.4s;\n    -webkit-border-radius: 50%;\n    -moz-border-radius: 50%;\n    -ms-border-radius: 50%;\n    -o-border-radius: 50%;\n  }\n\n  ", "\n\n  ",
    "\n"])), colors.highlight, colors.highlight, function (_a) {
    var label = _a.label;
    return label && "margin-left: 0.5rem;";
}, function (_a) {
    var checked = _a.checked;
    return checked &&
        "\n      background-color: " + colors.secondary + ";\n\n      :before {\n        opacity: 1;\n        transform: translateX(1.5rem);\n      }\n    ";
});
export var StyledToggleInput = styled.input(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  position: absolute;\n  opacity: 0;\n  z-index: 1;\n\n  // when focus, adds a border to its sibling StyledToggleSwitchSlider\n  :focus-visible + ", " {\n    outline: 0.125rem solid ", ";\n  }\n\n  :focus:not(:focus-visible) + ", " {\n    outline: none;\n  }\n"], ["\n  position: absolute;\n  opacity: 0;\n  z-index: 1;\n\n  // when focus, adds a border to its sibling StyledToggleSwitchSlider\n  :focus-visible + ", " {\n    outline: 0.125rem solid ", ";\n  }\n\n  :focus:not(:focus-visible) + ", " {\n    outline: none;\n  }\n"])), StyledToggleSwitchSlider, colors.highlight, StyledToggleSwitchSlider);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
