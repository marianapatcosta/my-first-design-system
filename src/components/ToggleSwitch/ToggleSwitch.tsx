import React, { ChangeEvent, CSSProperties } from 'react';
import {
  StyledToggleSwitch,
  StyledToggleSwitchLabel,
  StyledToggleSwitchSlider,
  StyledToggleInput,
} from './ToggleSwitch.style';

export interface ToggleSwitchProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
  /**
   * ToggleSwitch's label
   */
  label?: string;
  /**
   * label to display inside the ToggleSwitch, on the left side
   */
  leftLabel?: string;
  /**
   * label to display inside the ToggleSwitch, on the right side
   */
  rightLabel?: string;
  /**
   * ToggleSwitch's name
   */
  name?: string;
  /**
   * Is ToggleSwitch checked?
   */
  checked: boolean;
  /**
   * Is ToggleSwitch disabled?
   */
  disabled?: boolean;
  /**
   * Change handler
   */
  onChange: (event?: ChangeEvent<HTMLInputElement>) => void;
}

export const ToggleSwitch: React.FC<ToggleSwitchProps> = ({
  className,
  style,
  name,
  label,
  leftLabel,
  rightLabel,
  checked,
  disabled,
  onChange,
}) => {
  return (
    <StyledToggleSwitch className={className} style={style} disabled={disabled}>
      <StyledToggleInput
        type='checkbox'
        name={name || label}
        checked={checked}
        disabled={disabled}
        onChange={onChange}
      />
      {!!label && label}
      <StyledToggleSwitchSlider
        checked={checked}
        disabled={disabled}
        label={label}
        data-testid='toggle-slider'
      >
        <StyledToggleSwitchLabel disabled={disabled}>
          {checked && leftLabel}
        </StyledToggleSwitchLabel>
        <StyledToggleSwitchLabel>
          {!checked && rightLabel}
        </StyledToggleSwitchLabel>
      </StyledToggleSwitchSlider>
    </StyledToggleSwitch>
  );
};

ToggleSwitch.defaultProps = {
  id: '',
  className: '',
  style: {},
  name: '',
  label: '',
  leftLabel: '',
  rightLabel: '',
  checked: false,
  disabled: false,
  onChange: () => null,
};
