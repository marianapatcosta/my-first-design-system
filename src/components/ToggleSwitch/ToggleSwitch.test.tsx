import React from 'react';
import { render, getByText, fireEvent } from '@testing-library/react';
import { ToggleSwitch } from './ToggleSwitch';

describe('ToggleSwitch', () => {
  const defaultProps = {
    checked: false,
    leftLabel: 'off',
    rightLabel: 'on',
    onChange: () => null,
  };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<ToggleSwitch {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display label, if label prop is passed', () => {
    const { container } = render(
      <ToggleSwitch {...defaultProps} label='Toggle Switch' />
    );

    getByText(container, 'Toggle Switch');
  });

  test('should display leftLabel, if checked', () => {
    const { queryByText } = render(<ToggleSwitch {...defaultProps} checked={true}/>);

    expect(queryByText('off')).toBeTruthy();
    expect(queryByText('on')).toBeNull();
  });

  test('should display rightLabel, if not checked', () => {
    const { queryByText } = render(
      <ToggleSwitch {...defaultProps}  />
    );
    expect(queryByText('on')).toBeTruthy();
    expect(queryByText('off')).toBeNull();
  });

  test('should handle change events', () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <ToggleSwitch {...defaultProps} onChange={onChangeMock} />
    );
    const toggleSwitch = getByTestId('toggle-slider') as HTMLInputElement;
    fireEvent.click(toggleSwitch);

    expect(onChangeMock).toBeCalled();
  });

  test('should return null on change if onChange prop is not passed', () => {
    const result =
      ToggleSwitch.defaultProps?.onChange &&
      ToggleSwitch.defaultProps.onChange();
    expect(result).toBeNull();
  });

  describe('Disabled ToggleSwitch', () => {
    test('should be disabled', () => {
      const {
        container: { firstChild },
      } = render(<ToggleSwitch {...defaultProps} disabled />);

      expect(firstChild).toHaveAttribute('disabled');
    });

    test('should not handle change events', () => {
      const onChangeMock = jest.fn();
      const {
        container: { firstChild },
      } = render(
        <ToggleSwitch
          {...defaultProps}
          disabled={true}
          onChange={onChangeMock}
        />
      );

      firstChild && fireEvent.click(firstChild);

      expect(onChangeMock).not.toBeCalled();
    });
  });
});
