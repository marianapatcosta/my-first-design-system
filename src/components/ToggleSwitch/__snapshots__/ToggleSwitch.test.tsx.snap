// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch renders without errors and matches snapshot 1`] = `
<label
  class="sc-bdfBwQ LpMpT"
>
  <input
    class="sc-hKgILt jZJSDB"
    name=""
    type="checkbox"
  />
  <span
    class="sc-dlfnbm gaORok"
    data-testid="toggle-slider"
    label=""
  >
    <span
      class="sc-gsTCUz eFFOmb"
    />
    <span
      class="sc-gsTCUz eFFOmb"
    >
      on
    </span>
  </span>
</label>
`;
